### PR TITLE
feat: improve quote or rate discrimination

### LINKED ISSUE #8289: Receive address not showing on confirm and trade screen in trade modal for thorchain, 0x and portal trades
### Overview

When moving to the confirm and trade screen in the trade modal, the address the funds are going to does not populate straight away if its a thorchain, 0x or portals trade. you have to confirm and trade before the address appears. Lifi trade show the address straight away.

### References and additional details

image
image
https://jam.dev/c/3e7128fe-cc75-4d32-a050-bbc058342624 

### Acceptance Criteria

Should show receive address when move to preview trade

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

_No response_

### Sponsor / Stakeholder

_No response_

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR changes the discriminator between quotes and rates, i.e previously, `receiveAddress` was the big discriminator, which allowed us to detect a rate is a rate (non-executable, price only) vs. a quote (executable, with bip44/like params attached).

This had a few issues:

- Whenever the user *actually* connected a wallet, we were unable to do gas estimations if applicable 
- Since receiveAddress wouldn't be attached to a rate quote, it wouldn't be visible at confirm step (see #8289) until final quote time

This PR improves things by bringing quoteOrRate as a discriminator in trade quotes and trade rates, as well as consistently passing receiveAddress back.

## Issue (if applicable)

closes #8289

## Risk

> High Risk PRs Require 2 approvals

High. Touches the whole quotes vs. rates logic. Risks of broken swappers at rate time, final quotes being missing or being stale rates, as well as execution being either borked with infinite "Sign and broadcast" spin, or funds not landing on custom receive addy

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

- Test all swappers e2e without custom receive addy and ensure flow is happy e2e 
- Test all swappers e2e *with* custom receive addy (or with cross-account trades) and ensure flow is happy e2e 
- Test at least one sell Tx for each chain family (UTXO, EVM, ATOM, THOR, SOL)

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

- Same-Account EVM swaps

https://jam.dev/c/9eefdd34-fe4b-4fac-9a85-670a2484b22f

- Cross-Account EVM swaps

https://jam.dev/c/ea739b2e-a050-402e-bd89-fa9f6e0f7bb4

- Same-Account SOL swaps

https://jam.dev/c/6f1a586f-dbca-4e0f-a046-9884cb43d6ff

- Cross-account SOL swaps

Untestable yet as neither native nor Ledger supports Solana just yet

- Same-Account Cosmos swaps (cross-account irrelevant, there is only one *supported* token for Cosmos so ATOM sells are *always* cross-account)

https://jam.dev/c/937b5379-4994-417b-a30b-d8d374434b64

- UTXO swaps (cross-account irrelevant, there is only one token for UTXOs so UTXO sells are *always* cross-account)

https://jam.dev/c/014e03ba-ff6a-4877-9a23-adb8a587fbc4

- RUNE swaps (cross-account irrelevant, there is only one token for THORChain so RUNE sells are *always* cross-account

https://jam.dev/c/9a593c51-3ca2-46f2-8a0c-cd6e7dc21450

### CODE REVIEW

File: packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts
Comment (woodenfurniture) [inline packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts:18]: Unless my LSP is broken, this change doesnt seem to be required

**Review by woodenfurniture — COMMENTED** [review]
First pass, code wise looking good. Will wait for responses to feedback before completing full test.

File: packages/swapper/src/utils.ts
Comment (gomesalexandre) [reply packages/swapper/src/utils.ts:308]: re: first q, that would make a lot of sense actually, though not sure it makes sense *currently* as we are in the process of finding a way to make Li.Fi quotes/rate compliant and fetch final quotes using stepTransaction. Returning a `quote`, even though it's what we do, really doesn't fit the pattern
re: second one, actually cannot given our flow, since things referentially invalidate on wallety deps change for quotes to refetch, and you cannot continue to preview trade screen without a wallet connected

File: packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts:18]: tsc def. sad without!

image

File: packages/swapper/src/utils.ts
Comment (0xApotheosis) [reply packages/swapper/src/utils.ts:308]: Not super sure I understand this guy. My understanding is that anything with a `quoteOrRate` of `quote` will have a `receiveAddress`, and so the `return quote.quoteOrRate === 'quote'` part should be all that's necessary?

File: packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts
Comment (0xApotheosis) [reply packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts:18]: We do need the cast, but I don't think we need to go via `unknown`:

```suggestion
  })) as Result<LifiTradeRate[], SwapErrorRight>
```

File: packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts
Comment (0xApotheosis) [inline packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts:297]: We shouldn't need to cast as `unknown`.

```suggestion
  return Ok(promises.filter(isFulfilled).map(({ value }) => value)) as Result<
```

File: packages/swapper/src/utils.ts
Comment (gomesalexandre) [reply packages/swapper/src/utils.ts:308]: @0xApotheosis FYI, this change is gone as Li.Fi now follows the rates then quotes standard flow following 8309

Previously (i.e before I rebased this this mo') however, that wasn't the case, and `quoteOrRate` didn't cut it for Li.Fi:
Yes, `quoteOrRate` is indeed the new discriminator as part of this PR, but we previously never wanted to refetch a final quote for Li.Fi, meaning that a Li.Fi quote was fetched at rate time as a *rate* and should stay as such, but was actually executable, which is why there was a special case for Li.Fi.

Now that it follows our usual pattern, no need for it anymore!

Comment (gomesalexandre) [reply packages/swapper/src/utils.ts]: 44e0301847 and retested Li.Fi execution for sanity's sake:

https://jam.dev/c/6ae57343-9a9f-4ff3-a7ec-776ba8a9f63d

File: packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/LifiSwapper/getTradeRate/getTradeRate.ts]: Nice catch! Removed the `as unknown` cast in e52c6f7d6b and followed up with 214784c534

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts]: hell yeah 214784c534

File: packages/swapper/src/swappers/ThorchainSwapper/getThorTradeRate/getTradeRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/ThorchainSwapper/getThorTradeRate/getTradeRate.ts]: 100% - looks like there was an actual duplicate in types, it was already present in the base `ThorTradeQuoteSpecificMetadata`.

Removed dupe type and added one additional discriminator in  fa3e16798

Sanity retested both non-ROON sells and ROON sells:

https://jam.dev/c/0697e7ab-2218-4d6d-8f53-2934edf98b50

File: packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts]: https://github.com/shapeshift/web/commit/214784c534228d7742d9626a363c6f8b038a664f

File: packages/swapper/src/types.ts
Comment (gomesalexandre) [reply packages/swapper/src/types.ts:307]: Prefer to keep `quoteOrRate` as it is highly explicit, but open to naming changes in a follow-up!

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts:9]: Not applicable anymore, both trade quotes and trade rates *may* (in the case of trade rates, only with wallet connected) have BIP44ish params

File: packages/swapper/src/swappers/ChainflipSwapper/utils/getQuoteOrRate.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/utils/getQuoteOrRate.ts:283]: Removed inline type since TS would get confused with the push and think `(TradeQuote | TradeRate)[]` vs. `TradeQuote[] | TradeRate[]` and we'd end up with ugly casts

File: packages/swapper/src/types.ts
Comment (NeOMakinG) [reply packages/swapper/src/types.ts:314]: I also prefer quoteOrRate, a rate is not a quote type, it's a rate (in my mind at least)

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (NeOMakinG) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts]: @gomesalexandre it might be worth instead of duplicating to build an helper like `convertRateToQuote`, as we do for Solana instructions on chain adapter side, so you could easily `map(convertRateToQuote)`

File: packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts
Comment (NeOMakinG) [inline packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts:203]: We are not forced to explicitly set this to undefined now then, as it can be defined now

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx
Comment (NeOMakinG) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:1]: Do we want to remove the `receiveAddress` part from `shouldFetchTradeQuotes` and explicitly use `quoteOrRate` now?

**Review by NeOMakinG — COMMENTED** [review]
A few small questions/comments but it's pretty straightforward, looking good an way better than using `receiveAddress` as a discriminator

Testing at runtime right now

File: packages/swapper/src/types.ts
Comment (gomesalexandre) [reply packages/swapper/src/types.ts:314]: 1 rate = 1 rate

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts]: Would prefer to avoid it as this would validate the current way, however, the current way is far from optimal and I would prefer to spend time actually improving types to avoid casts, rather than adding bandaid utils 🙏🏽

File: packages/swapper/src/swappers/ChainflipSwapper/utils/getQuoteOrRate.ts
Comment (gomesalexandre) [reply packages/swapper/src/swappers/ChainflipSwapper/utils/getQuoteOrRate.ts:342]: Casting is not required anymore as I've removed the strong typings of `ratesOrQuotes` (initialised as `any`, type inferred thereafter), but would prefer to keep it for sanity's sake.

Actually removed `TradeQuoteOrRate` in 558a016650 - we weren't using it except in the one place, and we should be explicit in most cases (Chainflip is the exception following https://github.com/shapeshift/web/pull/8152, unfortunately bad timing as David started while this was still in-flight, and the previous split between quotes and rates without a shared util made a lot more sense high-level)

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx
Comment (NeOMakinG) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:1]: ignore this comment, it's not true anymore regarding this PR

Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:1]: ` isExecutableTradeQuote` is already doing precisely that: 
https://github.com/shapeshift/web/blob/558a0166502e9214c0a8b54813b0135d403d5e56/packages/swapper/src/utils.ts#L304-L305

This is now paranoia at this point, but prefer to keep said paranoia since we already have it.

**Review by NeOMakinG — APPROVED** [review]
BTC => Rune
Inbound: https://viewblock.io/thorchain/tx/00b484ba4c58a66d0d17e1578153864ba4b79961cf3e17bbd457d8385c7cb38c
https://jam.dev/c/b4fba010-284a-4340-bfa2-d828fc759935

AVAX to ETH on thorchain
https://jam.dev/c/97d64b09-20a6-40ce-ad1a-7f2f3611b5c1

EVM Cross account lifi
https://jam.dev/c/7dbb80a3-0a15-4af1-b413-28760a93fe53

EVM not cross account 0x
https://jam.dev/c/7dbb80a3-0a15-4af1-b413-28760a93fe53

I had some flaky behaviors that are looking bad upstream I think:

Chainflip swap "error c"
https://jam.dev/c/5578b075-f9cf-4631-9ea9-d9ae741f5e13

Rune to cosmos swap Invalid pubkey
https://jam.dev/c/ea135569-1477-42f5-89b3-395a1e725b2d

Cosmos to rune invalid pubkey
https://jam.dev/c/eb224475-6e14-446e-939f-746ed2d73e12

RUNE => Avax (Invalid pubkey)
https://jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509 

LiFi ❌  Route error is not handled gracefully
https://jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509

EVM cross account CowSwap ❌
https://jam.dev/c/f9e0ede7-0b02-4f8d-8015-0a86b552f081 

I think it's not related to the PR, I can approve it as we need a second pair of eyes anyway, but can se double check those behaviors to ensure we don't implement any regression? From a code perspective, I don't how we would introduce so much things

**Comment by gomesalexandre** [comment]
> I had some flaky behaviors that are looking bad upstream I think:

@NeOMakinG confirming all of the issues above are either expected (fixed in different PRs/captured as issues previously) and unrelated, or fixed in the case of Chainflip swaps.

> Chainflip swap "error c" [jam.dev/c/5578b075-f9cf-4631-9ea9-d9ae741f5e13](https://jam.dev/c/5578b075-f9cf-4631-9ea9-d9ae741f5e13)

I was able to repro:

image

develop is also borked however, this never loads:

image

Fixed in bcd115106b for Chainflip and Li.Fi: https://jam.dev/c/524d3b1c-d0f1-4fcc-8a39-ad6609ffc6d4

> 
> Rune to cosmos swap Invalid pubkey [jam.dev/c/ea135569-1477-42f5-89b3-395a1e725b2d](https://jam.dev/c/ea135569-1477-42f5-89b3-395a1e725b2d)

Unrelated to this PR, see https://github.com/shapeshift/web/issues/8351

> Cosmos to rune invalid pubkey [jam.dev/c/eb224475-6e14-446e-939f-746ed2d73e12](https://jam.dev/c/eb224475-6e14-446e-939f-746ed2d73e12)

Unrelated to this PR, see https://github.com/shapeshift/web/issues/8351


> RUNE => Avax (Invalid pubkey) [jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509](https://jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509)

Unrelated to this PR, see https://github.com/shapeshift/web/issues/8351

> 
> LiFi ❌ Route error is not handled gracefully [jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509](https://jam.dev/c/e1fceca8-5826-487c-b87c-47ecb8792509)

Looks like this isn't the correct Jam as I there's no Li.Fi in it, but Li.Fi errors are unrelated to this PR, bridges fixed in https://github.com/shapeshift/web/pull/8348 (though that PR didn't make its way into develop yet, so will still happen here) and Li.Fi unstable quotes are tackled in a PR on top of this one (meaning we may still encounter it here, though keep it mind it's a v. rare occurence) i.e https://github.com/shapeshift/web/pull/8350

> 
> EVM cross account CowSwap ❌ [jam.dev/c/f9e0ede7-0b02-4f8d-8015-0a86b552f081](https://jam.dev/c/f9e0ede7-0b02-4f8d-8015-0a86b552f081)

@NeOMakinG not sure what we mean by cross-account here, as CoW doesn't support cross-account trades (see in the Jam below), however, confirmed CoW is happy:

https://jam.dev/c/15a5d2eb-9f37-4afd-9307-9944d45394fa

**Review by 0xApotheosis — APPROVED** [review]
Looking good now ser.

I was a little blocked on some testing steps, but could confirm:

✅ Same chain EVM
✅ Cross-chain EVM
❔ UTXO - THORChain node down at time of writing
❔ Solana - I couldn't get any SOL on my Phantom wallet due to lack of test funds

**Comment by gomesalexandre** [comment]
Tyvm for the review big @0xApotheosis!

> ❔ UTXO - THORChain node down at time of writing
> ❔ Solana - I couldn't get any SOL on my Phantom wallet due to lack of test funds

Tested Jupiter for sanitys's sake so you don't have to:

https://jam.dev/c/c3c5cd99-a1ec-461f-a904-d73698b1981e

**Pull request merged**
